File: JANL/Types/FixedQueue.vb
Namespace Types

	Public Class FixedQueue(Of T)
		Inherits Queue(Of T)
		Private Limit As Integer

		Public Sub New(_limit As Integer)
			MyBase.New(_limit)
			Limit = _limit
		End Sub

		Public Overloads Sub Enqueue(item As T)
			While (Count >= Limit)
				Dequeue()
			End While
			MyBase.Enqueue(item)
		End Sub

	End Class
End Namespace

File: JANL/Generic/FixedQueue.cs
using System.Collections.Generic;

namespace JANL.Generic
{
    /// <summary>
    /// Очередь фиксированного размера
    /// </summary>
    /// <typeparam name="T">Тип очереди</typeparam>
    public class FixedQueue<T> : Queue<T>
    {
        /// <summary>
        /// Создает новый экземпляр очереди
        /// </summary>
        /// <param name="limit">Размер очереди</param>
        public FixedQueue(int limit) : base(limit) { _limit = limit; }

        /// <summary>
        /// Добавляет элемент в конец очереди
        /// </summary>
        /// <param name="item">Элемент</param>
        public new void Enqueue(T item)
        {
            Truncate();
            base.Enqueue(item);
        }

        private void Truncate()
        {
            while (Count >= Limit) { Dequeue(); }
        }

        #region Properties
        private int _limit;

        /// <summary>
        /// Размер очереди
        /// </summary>
        public int Limit
        {
            get => _limit;
            set
            {
                if (value == _limit) { return; }
                _limit = value;
                Truncate();
            }
        }

        #endregion Properties
    }
}